asllib/doc/TopLevel.tex

```diff
@@ -0,0 +1,104 @@
+    \item \Prosebuildast{$\vstdparse$}{$\vstdast$\ProseTerminateAs{\BuildErrorConfig}};
+    \item define $\vstdasbuiltin$ by applying $\setbuiltin$ to each top-level declaration in\\
+          $\vstdast$\ProseOrTypeError;
+    %%%%%
+    \item \Proseaslscan{$\vspectext$}{$\spectoken$}{$\vspectokens$\ProseTerminateAs{\LexicalError}};
+    \item \Proseaslparse{$\vspectokens$}{$\vspecparse$\ProseTerminateAs{\ParseError}};
+    \item \Prosebuildast{$\vspecparse$}{$\vspecast$\ProseTerminateAs{\BuildErrorConfig}};
+    %%%%%
+    \item define $\vuntypedast$ as the concatenation of the AST $\vstdasbuiltin$ and
+          the AST $\vspecast$ (both are lists of $\decl$);
+    \item \Prosetypecheckast{$\vuntypedast$}{empty static global environment}{$\vtypedast$}{$\tenv$\ProseOrTypeError};
+    \item \Proseevalspec{$\vtypedast$}{$\tenv$}{the integer $\vret$ and the execution graph $\vg$\ProseOrError}.
+\end{itemize}
+
+\subsubsection{Formally}
+\begin{mathpar}
+\inferrule{
+    \aslscan(\spectoken, \vstdtext) \scanarrow \vstdtokens \terminateas \LexicalError\\\\
+    \aslparse(\vstdtokens) \parsearrow \vstdparse \terminateas \ParseError\\\\
+    \buildast(\vstdparse) \astarrow \vstdast \terminateas \BuildErrorConfig\\\\
+    \vi \in \listrange(\vstdast): \setbuiltin(\vstdast_i) \typearrow \vstddeclast_i \OrTypeError\\\\
+    \vstdasbuiltin \eqdef [\vi \in \listrange(\vstdast): \vstddeclast_i]\\
+    \aslscan(\spectoken, \vspectext) \scanarrow \vspectokens \terminateas \LexicalError\\\\
+    \aslparse(\vspectokens) \parsearrow \vspecparse \terminateas \ParseError\\\\
+    \buildast(\vspecparse) \astarrow \vstdast \terminateas \BuildErrorConfig\\\\
+    \vuntypedast \eqdef \vstdasbuiltin \concat \vspecast\\
+    \typecheckast(G^{\emptytenv}, \vuntypedast) \typearrow (\vtypedast, \tenv) \OrTypeError\\\\
+    \evalspec(\tenv, \vtypedast) \evalarrow (\nvint(\vret), \vg) \OrDynError
+}{
+    \checkandinterpret(\vspectext, \vstdtext) \longrightarrow (\nvint(\vret), \vg)
+}
+\end{mathpar}
+
+\subsubsection{TypingRule.SetBuiltin\label{sec:TypingRule.SetBuiltin}}
+The helper function
+\hypertarget{def-setbuiltin}{}
+\[
+  \setbuiltin(\overname{\decl}{\vdecl}) \aslto
+  \overname{\decl}{\vdeclp}
+  \cup \overname{\TTypeError}{\TypeErrorConfig}
+\]
+sets the $\funcbuiltin$ flag of a top-level function declaration, which is used to identify standard library functions in \secref{TypingRule.InsertStdlibParam}.
+It produces a type error when given a top-level declaration which is not a function.
+
+\subsubsection{Prose}
+All of the following apply:
+\begin{itemize}
+  \item \Prosechecktrans{that $\vdecl$ is a function}{\BuiltinExpectedToBeFunction};
+  \item define $\funcsigp$ as $\funcsig$ with its $\funcbuiltin$ flag set to $\True$;
+  \item define $\vdeclp$ as $\DFunc(\funcsigp)$.
+\end{itemize}
+
+\subsubsection{Formally}
+\begin{mathpar}
+\inferrule{
+  \checktrans{\astlabel(\vdecl) = \DFunc}{\BuiltinExpectedToBeFunction} \typearrow \True \OrTypeError\\\\
```

Comment:
nit: the type error here can be more specific now (unless I am mistaken)

Reply:
Right, but I've used the same template where the error code appears in the `\checktrans` second argument and the error configuration in the outcome is just the generic one. I should probably fix this globally at some point.